MathLib/mechanics_invariants.h

```diff
@@ -0,0 +1,166 @@
+    static Eigen::Matrix<double, KelvinVectorSize, KelvinVectorSize> const
+        P_sph;
+    /// Kelvin mapping of 2nd order identity
+    static Eigen::Matrix<double, KelvinVectorSize, 1> const ivec;
+
+    // Determinant of a matrix in Kelvin vector representation.
+    static double determinant(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& vec);
+
+    // Inverse of a matrix in Kelvin vector representation.
+    static Eigen::Matrix<double, KelvinVectorSize, 1> inverse(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& vec);
+
+    static double effectiveStress(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return std::sqrt(3 * J2(deviatoric_stress));
+    }
+
+    static double J2(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return 0.5 * deviatoric_stress.transpose() * deviatoric_stress;
+    }
+
+    static double J3(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return determinant(deviatoric_stress);
+    }
+
+    static double I1(Eigen::Matrix<double, KelvinVectorSize, 1> const& stress)
+    {
+        return stress.template block<3, 1>(0, 0).sum();
+    }
+};
+
+template <>
+inline double Invariants<6>::determinant(Eigen::Matrix<double, 6, 1> const& vec)
+{
+    return vec(0) * vec(1) * vec(2) + vec(3) * vec(4) * vec(5) / std::sqrt(2.) -
+           vec(3) * vec(3) * vec(2) / 2. - vec(4) * vec(4) * vec(0) / 2. -
+           vec(5) * vec(5) * vec(1) / 2.;
+}
+
+template <>
+inline double Invariants<4>::determinant(Eigen::Matrix<double, 4, 1> const& vec)
+{
+    return vec(0) * vec(1) - vec(3) * vec(3) / 2.;
+}
+
+template <>
+inline double Invariants<1>::determinant(Eigen::Matrix<double, 1, 1> const& vec)
+{
+    return vec(0);
+}
+
+template <>
+inline Eigen::Matrix<double, 1, 1> Invariants<1>::inverse(
+    Eigen::Matrix<double, 1, 1> const& vec)
+{
+    Eigen::Matrix<double, 1, 1> inv;
+    inv(0) = 1 / vec(0);
+    return inv;
+}
+
+template <>
+inline Eigen::Matrix<double, 4, 1> Invariants<4>::inverse(
+    Eigen::Matrix<double, 4, 1> const& vec)
+{
+    Eigen::Matrix<double, 4, 1> inv;
+    inv(0) = vec(1);
+    inv(1) = vec(0);
+    inv(2) = 0;
+    inv(3) = - vec(3);
+    return inv / determinant(vec);
+}
+
+template <>
+inline Eigen::Matrix<double, 6, 1> Invariants<6>::inverse(
+    Eigen::Matrix<double, 6, 1> const& vec)
+{
+    Eigen::Matrix<double, 6, 1> inv;
+    inv(0) = vec(1) * vec(2) - vec(4) * vec(4) / 2.;
+    inv(1) = vec(0) * vec(2) - vec(5) * vec(5) / 2.;
+    inv(2) = vec(0) * vec(1) - vec(3) * vec(3) / 2.;
+    inv(3) = vec(4) * vec(5) / std::sqrt(2.) - vec(3) * vec(2);
+    inv(4) = vec(3) * vec(5) / std::sqrt(2.) - vec(4) * vec(0);
+    inv(5) = vec(4) * vec(3) / std::sqrt(2.) - vec(1) * vec(5);
+    return inv / determinant(vec);
+}
+
+template <int KelvinVectorSize>
+Eigen::Matrix<double, KelvinVectorSize, KelvinVectorSize> initP_dev()
```

Comment:
`detail` ns not possible, because specialization must be in the same ns as the declaration.
Moving to impl (or even cpp) after fixing the other things to reduce confusion.